src/components/application_manager/src/resumption/resumption_data_processor.cc

```diff
@@ -0,0 +1,1125 @@
+                               << has_data_to_restore);
+
+  return has_data_to_restore;
+}
+
+bool ResumptionDataProcessor::HasGlobalPropertiesToRestore(
+    const smart_objects::SmartObject& saved_app) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  const smart_objects::SmartObject& global_properties =
+      saved_app[strings::application_global_properties];
+
+  const bool has_gl_props_to_restore =
+      !global_properties[strings::help_prompt].empty() ||
+      !global_properties[strings::keyboard_properties].empty() ||
+      !global_properties[strings::menu_icon].empty() ||
+      !global_properties[strings::menu_title].empty() ||
+      !global_properties[strings::timeout_prompt].empty() ||
+      !global_properties[strings::vr_help].empty() ||
+      !global_properties[strings::vr_help_title].empty();
+
+  LOG4CXX_DEBUG(logger_,
+                std::boolalpha
+                    << "Application has global properties to restore: "
+                    << has_gl_props_to_restore);
+
+  return has_gl_props_to_restore;
+}
+
+bool ResumptionDataProcessor::HasSubscriptionsToRestore(
+    const smart_objects::SmartObject& saved_app) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  const smart_objects::SmartObject& subscriptions =
+      saved_app[strings::application_subscriptions];
+
+  const bool has_ivi_subscriptions =
+      !subscriptions[strings::application_vehicle_info].empty();
+
+  const bool has_button_subscriptions =
+      !(subscriptions[strings::application_buttons].length() == 1 &&
+        static_cast<hmi_apis::Common_ButtonName::eType>(
+            subscriptions[strings::application_buttons][0].asInt()) ==
+            hmi_apis::Common_ButtonName::CUSTOM_BUTTON);
+
+  const bool has_waypoints_subscriptions =
+      subscriptions[strings::subscribed_for_way_points].asBool();
+
+  const bool has_appservice_subscriptions =
+      subscriptions.keyExists(app_mngr::hmi_interface::app_service) &&
+      !subscriptions[app_mngr::hmi_interface::app_service].empty();
+
+  const bool has_system_capability_subscriptions =
+      subscriptions.keyExists(strings::system_capability) &&
+      !subscriptions[strings::system_capability].empty();
+
+  const bool has_subscriptions_to_restore =
+      has_ivi_subscriptions || has_button_subscriptions ||
+      has_waypoints_subscriptions || has_appservice_subscriptions ||
+      has_system_capability_subscriptions;
+
+  LOG4CXX_DEBUG(logger_,
+                std::boolalpha << "Application has subscriptions to restore: "
+                               << has_subscriptions_to_restore);
+
+  return has_subscriptions_to_restore;
+}
+
+utils::Optional<uint32_t> ResumptionDataProcessor::GetAppIdWaitingForResponse(
+    const hmi_apis::FunctionID::eType function_id, const int32_t corr_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  auto predicate =
+      [function_id,
+       corr_id](const std::pair<ResumptionRequestID, std::uint32_t>& item) {
+        return item.first.function_id == function_id &&
```

Comment:
@iCollin well, I would say this have to be done first of all for safety.
We know internally, that SDL generates unique correlation id by simple increment, but if `max_correlation_id` setting is too small then it is possible to have two requests with the same correlation id generated. This may confuse SDL.
Another potential vulnerability is related to HMI. Again, we know that HMI is implemented internally to respond with the same correlation id, however by some mistake it may send response for different function_id with the same correlation_id which will also confuse SDL.
So to make SDL more protected, I think it is better to keep this check. It does not impact performance anyway.